File: Data Structures & Algorithms/binary-tree-diameter/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    constructor() {
        this.diameter = 0
    }

    diameterOfBinaryTree(root) {
        this.updateDiameterOfTree(root)
        return this.diameter
    }


    updateDiameterOfTree(root) {
        let heightLeft = 0, heightRight = 0;

        if (!root) {
            return 0
        }

        if (root.left) {
            heightLeft = 1 + this.updateDiameterOfTree(root.left)
        }

        if (root.right) {
            heightRight = 1 + this.updateDiameterOfTree(root.right)
        }

        this.diameter = Math.max(this.diameter, heightLeft + heightRight)

        return Math.max(heightLeft, heightRight)
    }
}

File: Data Structures & Algorithms/convert-bst-to-greater-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode convertBST(TreeNode root) {
        int sum = 0;

        _convertBST(root, sum);

        return root;
    }

    private int _convertBST(TreeNode root, int sum) {
        if (root == null) return sum;
        sum = _convertBST(root.right, sum);

        sum = sum + root.val;
        root.val = sum;
        sum = _convertBST(root.left, sum);

        return sum;
    }
}